### PR TITLE
Fix multi callback

### PR DESCRIPTION
**Pull request by TimWSpence**
Allow registering multiple callbacks with the same name/labels. Also implement reference counting to solve lifecycle issues when multiple scopes open the same metric resource.

Addresses https://github.com/permutive-engineering/prometheus4cats/issues/38.

### CODE REVIEW

File: testkit/src/main/scala/prometheus4cats/testkit/CallbackRegistrySuite.scala [hunk 1]
```diff
@@ -424,4 +426,118 @@ trait CallbackRegistrySuite[State] extends RegistrySuite[State] { self: CatsEffe
         }
     }
   }
+
```
Comment (janstenpickle) [inline testkit/src/main/scala/prometheus4cats/testkit/CallbackRegistrySuite.scala:429]: I _think_ this is missing a case where you have a callback of the same name with different label names. This _should_ be permissable now, thanks to the hacking 😂

File: java/src/main/scala/prometheus4cats/javasimpleclient/JavaMetricRegistry.scala [hunk 2]
```diff
@@ -502,31 +539,19 @@ class JavaMetricRegistry[F[_]: Async: Logger] private (
...
-            val metrics = nel.map { case (value, labels) =>
-              makeLabelledFamily(stringName, commonLabelNames, (f(labels) ++ commonLabelValues).asJava, value)
-            }
-
-            metrics.toList.asJava
-          case None => List.empty[Collector.MetricFamilySamples].asJava
-        }
-
-    }
+    val x: F[NonEmptyList[MetricFamilySamples]] = callback.map(_.map { case (value, labels) =>
```
[the remark this answers is not in the document]
Comment (TimWSpence) [reply java/src/main/scala/prometheus4cats/javasimpleclient/JavaMetricRegistry.scala:550]: 😂 yup

File: java/src/main/scala/prometheus4cats/javasimpleclient/JavaMetricRegistry.scala [hunk 3]
```diff
@@ -502,31 +539,19 @@ class JavaMetricRegistry[F[_]: Async: Logger] private (
...
+      makeLabelledFamily(stringName, commonLabelNames, (f(labels) ++ commonLabelValues).toList, value)
+    })
 
-    registerCallback(metricType, prefix, name, labelNames ++ commonLabels.value.keys.toIndexedSeq, collector)
+    registerCallback2(metricType, prefix, name, x)
```
[the remark this answers is not in the document]
Comment (TimWSpence) [reply java/src/main/scala/prometheus4cats/javasimpleclient/JavaMetricRegistry.scala:554]: 😂 fair point!

File: testkit/src/main/scala/prometheus4cats/testkit/CallbackRegistrySuite.scala (hunk 1 above)
Comment (TimWSpence) [reply testkit/src/main/scala/prometheus4cats/testkit/CallbackRegistrySuite.scala:429]: Yeah, I wondered about that and then forgot about it again 🤦 Good spot 👍

File: java/src/main/scala/prometheus4cats/javasimpleclient/JavaMetricRegistry.scala [hunk 4]
```diff
@@ -50,12 +47,13 @@ import scala.concurrent.duration._
 import scala.jdk.CollectionConverters._
 
 class JavaMetricRegistry[F[_]: Async: Logger] private (
```
[the remark this answers is not in the document]
Comment (TimWSpence) [reply java/src/main/scala/prometheus4cats/javasimpleclient/JavaMetricRegistry.scala:49]: You mean so there isn't one type that implements both kinds of registry? Or just move the implementation into an abstract class and have `JavaMetricRegistry` extend it?

File: testkit/src/main/scala/prometheus4cats/testkit/CallbackRegistrySuite.scala [hunk 5]
```diff
@@ -540,4 +540,52 @@ trait CallbackRegistrySuite[State] extends RegistrySuite[State] { self: CatsEffe
     }
   }
 
+  test("allows building a callback with different labels but the same name") {
```
Comment (janstenpickle) [inline testkit/src/main/scala/prometheus4cats/testkit/CallbackRegistrySuite.scala:543]: Could you add one with differing numbers of labels as well please?

**Pull request merged**